clients/drcachesim/reader/config_reader.cpp

```diff
@@ -0,0 +1,440 @@
+               core_inst_caches[cache->core]++;
+            }
+            if (cache->type == "data" || cache->type == "unified") {
+               core_data_caches[cache->core]++;
+            }
+        }
+
+        // Associate a cache with its parent and children caches.
+        if (cache->parent != MEMORY) {
+            auto parent_it = caches_map.find(cache->parent);
+            if (parent_it != caches_map.end()) {
+                auto &parent = parent_it->second;
+                // Check that the cache types are compatible.
+                if (parent->type != "unified" &&
+                    cache->type != parent->type) {
+                    ERRMSG("Cache %s and its parent have incompatible types\n",
+                           cache_name.c_str());
+                    return false;
+                }
+
+                // Add the cache to its parents children.
+                parent->children.push_back(cache_name);
+            }
+            else {
+                ERRMSG("Cache %s has a listed parent %s that does not exist\n",
+                       cache_name.c_str(), cache->parent.c_str());
+                return false;
+            }
+
+            // Check for cycles between the cache and its parent.
+            string parent = cache->parent;
+            while (parent != MEMORY) {
+                if (parent == cache_name) {
+                    ERRMSG("Cache %s and its parent %s have a cyclic reference\n",
+                           cache_name.c_str(), cache->parent.c_str());
+                    return false;
+                }
+                parent = caches_map.find(parent)->second->parent;
+            }
+        }
+    }
+
+    // Check that each core has exactly one instruction and one data caches or
+    // exactly one unified cache.
+    for (int i = 0; i < num_cores; i++) {
+        if (core_inst_caches[i] != 1) {
+            ERRMSG("Core %d has %d instruction caches. Must have exactly 1\n",
+                   i, core_inst_caches[i]);
+            return false;
+        }
+        if (core_data_caches[i] != 1) {
+            ERRMSG("Core %d has %d data caches. Must have exactly 1\n",
+                   i, core_data_caches[i]);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool
+config_reader_t::convert_string_to_size(const string &s, uint64_t &size) {
+    char suffix = *s.rbegin();  // s.back() only in C++11
+    int scale;
+    switch (suffix) {
+        case 'K':
+        case 'k': scale = 1024; break;
+        case 'M':
+        case 'm': scale = 1024*1024; break;
+        case 'G':
+        case 'g': scale = 1024*1024*1024; break;
+        default: scale = 1;
+    }
+
+    std::string toparse = s;
+    if (scale > 1)
+        toparse = s.substr(0, s.size()-1); // s.pop_back() only in C++11
+
+    // While the overall size is likely too large to be represented
+    // by a 32-bit integer, the prefix number is usually not.
```

Comment:
Hmm, so this is an exact duplicate of droption_t<bytesize_t>::convert_from_string -- how about an XXX comment up top  about sharing the code somehow

Reply:
Done.